---
title: "Silberstein- stem cell screening"
output:
  html_document:
    theme: cosmo
    toc: true
    toc_depth: 4
    fig_caption: true
    fig_width: 8
    fig_height: 6
author: "Meeta Mistry"
---

```{r setup, echo=FALSE}
 
# Setup report details
clientname="Myriam Armant"
clientemail="myriam.armant@childrens.harvard.edu"
lablocation="HSCI"
analystname="Meeta Mistry"
analystemail="mmistry@hsph.harvard.edu"
```

RNA-Seq analysis for `r clientname` (`r clientemail`) at `r lablocation`. Contact `r analystname` (`r analystemail`) for additional details. Request from client was:     

Looking for concordance between microarray and RNASeq data. Dataset is composed of paired samples from four donors to evaluate the expression at two stages of limbal stem cell cultivation (treatment for LSCD):                        

1. P0: secondary cell cultures on plastic dish
2. AM: grafts released from plastic onto amniotic membrane


## Workflow:   
  * run BAM files through [bcbio](https://bcbio-nextgen.readthedocs.org/en/latest/index.html)
  * [assess QC](./partek_aligned/qc-summary.html) 
  * simple DEG list from Microarray data versus RNA-Seq
  * for microarray analysis some help from [github](https://github.com/benilton/oligo/wiki/Getting-the-grips-with-the-oligo-Package)

## Setup

### Bioconductor and R libraries used

```{r libraries, echo=TRUE}
loadlibs <- function(){
library(ggplot2)
library(reshape)
library(gplots)
library(arrayQualityMetrics)
library(RColorBrewer)
library(oligo)
library(CHBUtils)
library(ggdendro)
library(grid)
library(gridExtra)
library(limma)
library(edgeR)
library(DESeq2)
library(Biobase)
library(genefilter)
library(splitstackshape)
}
suppressPackageStartupMessages(loadlibs())
```

### Set variables
```{r directories, echo=TRUE}
baseDir=getwd()
resultsDir=paste(baseDir, "/results", sep="")

heatcolors.1 <- rev(brewer.pal(6, "YlOrRd"))
heatcolors.2 <- colorRampPalette(brewer.pal(9, "GnBu"))(100)
cbPalette <- cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", 
                            "#D55E00", "#CC79A7", "#000000")
```

## Microarray
### Load data
```{r load data, warning=FALSE, message=FALSE}

# Load CEL files
array.data <- read.celfiles(filenames=list.celfiles('./microarray', full.name=TRUE), verbose=FALSE)

# Background correct aand normalize
rma.data <- rma(array.data, target="core")

# Add pheno data
covars <- read.delim(file.path(baseDir, 'microarray/covars.desc'), row.names=1) 
colnames(exprs(rma.data)) <- row.names(covars)
pData(rma.data) <- covars
pData(rma.data)$BAM_filenames <- NULL
pData(rma.data)$Donors <- factor(pData(rma.data)$Donors)

# Add annotation data
library(annotate)
library(hugene10sttranscriptcluster.db)
annodb <- "hugene10sttranscriptcluster.db"
ID     <- featureNames(rma.data)
Symbol <- as.character(lookUp(ID, annodb, "SYMBOL"))
Ensembl   <- as.character(lookUp(ID, annodb, "ENSEMBL"))
Entrez <- as.character(lookUp(ID, annodb, "ENTREZID"))
ann <- cbind(ID, Symbol, Ensembl, Entrez)

fData(rma.data) <- data.frame(cbind(Symbol, Ensembl, Entrez), row.names=ID)
```

### QC Analysis
The report for the raw data can be found here: [Raw QC](./results/report_raw/index.html). The report for normalized data can be found here: [RMA QC](./results/report_rma/index.html) According to the boxplot of intensity distributions in both the raw data and processed data sample 4B appears to be an outlier. Although from the PCA and dendrogram based on Euclidian distance - it seems more likely that Sample 5A is the outlier (figures below).

```{r QC_report, echo=FALSE, eval=FALSE}
# Raw data
arrayQualityMetrics(expressionset=data, intgroup='Stage',
                     outdir='./results/report_raw',
                     force=TRUE,
                    do.logtransform=TRUE)
# Normalized data
 arrayQualityMetrics(expressionset=rma.data, intgroup='Stage',
                     outdir='./results/report_rma',
                     force=TRUE,
                    do.logtransform=FALSE)
```

```{r qcarray, echo=FALSE, fig.align='center', fig.width=18, warning=FALSE, message=FALSE}
# PCA plot of first and second PCs
p1 <- mds(exprs(rma.data), condition=pData(rma.data)$Stage , k=length(colnames(rma.data)) - 1)


# Dendrogram
  meta.x <- pData(rma.data)
  myDist <- dist(t(exprs(rma.data)))
  myTree <-hclust(myDist)
  dhc <- as.dendrogram(myTree)
  ddata <- dendro_data(dhc, type="rectangle")
  ddata$labels <- merge(ddata$labels, meta.x, by.x="label", by.y="row.names")
p2 <- ggplot(segment(ddata)) +
    geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) +
    theme_dendro() +
    geom_text(data=label(ddata), aes(x=x, y=y, label=label, color= label(ddata)[['Stage']], 
                                     hjust=-0.1), size=6) +
    coord_flip() + scale_y_reverse(expand=c(0.2, 50)) +
    theme(axis.text.x=element_blank(),
          axis.text.y=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          legend.title=element_blank())

grid.arrange(p1, p2, ncol=2)
```
 
### Differential expression
Looking at expression differences between the two stages P0 and AM we find zero genes that are differntially expressed. The p-value distribution indicates there is a high frequency of low p-values, but at an FDR < 0.1 only six genes are identified and listed in the table below.
```{r limma paired design, echo=TRUE, fig.align='center', results='asis'}

# Remove outlier sample
remove <- which(colnames(rma.data) == "MA_B3_5A.CEL")
exprs(rma.data) <- exprs(rma.data)[,-remove]
pData(rma.data) <- pData(rma.data)[-remove,]

# Get model matrix
mod <- model.matrix(~Stage + Donors, pData(rma.data))

# Fit a linear model
fit<-lmFit(rma.data, mod)
fit2<-eBayes(fit)

# Set threshold 
p.cutoff <- 0.1
logfc.cutoff <- 2

gene_list <- topTable(fit2, coef=2, number=nrow(exprs(rma.data)), sort.by="logFC")

# P-value Distribution
hist(gene_list$P.Value, col="darkgrey", border=F, xlab="P-value", main="P-value Distribution")

# Significant genes
sigGenes <- gene_list[which(gene_list$adj.P.Val < p.cutoff), c('Symbol', 'logFC', 'P.Value', 'adj.P.Val')]
sigGenes$P.Value <- formatC(sigGenes$P.Value, digits=3, format="e")
kable(sigGenes, format='markdown', row.names=F)
```

## RNA-Seq (IonTorrent)
### Load data

```{r loadcounts}
seq.data <- read.delim(file.path(baseDir, 'partek_aligned/htseq-count/combined.counts'), header=T, sep="\t",
                       row.names=1, as.is=T) 
seq.meta <- read.delim(file.path(baseDir, 'partek_aligned/covars.desc'), header=T, sep="\t")
seq.meta <- seq.meta[which(rownames(seq.meta) %in% colnames(seq.data)),]
seq.meta$BAM_filenames <- NULL
seq.meta$Donors <- droplevels(factor(seq.meta$Donors))
```

### Quality control
Using the aligned BAM files from Partek Flow Server, several [summary plots](./partek_aligned/qc-summary.html) were used to evaluate the sample quality. We found that while almost half of the samples had a low mapping rate ( < 60%), the number of genes detected was roughly equal across all samples (~20,000).        

We used [ht-seq](http://www-huber.embl.de/users/anders/HTSeq/doc/count.html) to generate a matrix of counts (mapped reads) for each feature (gene). Similar to the microarray data, the dendrogram below shows that samples assayed by IonTorrent sequencing cluster by Stage with the exception of one sample (AM_donor4). 

```{r qcseq, echo=FALSE, fig.align='center', fig.width=22, warning=FALSE, message=FALSE}
# PCA plot of first and second PCs
dge <- DGEList(counts=seq.data)
dge <- calcNormFactors(dge)
normalized_counts = cpm(dge, normalized.lib.sizes=TRUE)
p1 <- mds(normalized_counts, condition=seq.meta$Stage, k=length(colnames(seq.data)) - 1)

# Dendrogram
  meta.x <- seq.meta
  myDist <- dist(t(normalized_counts))
  myTree <-hclust(myDist)
  dhc <- as.dendrogram(myTree)
  ddata <- dendro_data(dhc, type="rectangle")
  ddata$labels <- merge(ddata$labels, meta.x, by.x="label", by.y="row.names")
p2 <- ggplot(segment(ddata)) +
    geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) +
    theme_dendro() +
    geom_text(data=label(ddata), aes(x=x, y=y, label=label, color= label(ddata)[['Stage']], 
                                     hjust=-0.1), size=6) +
    coord_flip() + scale_y_reverse(expand=c(0.2, 50)) +
    theme(axis.text.x=element_blank(),
          axis.text.y=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          legend.title=element_blank())

grid.arrange(p1, p2, ncol=2)
```


### Differential expression
Using two methods (DESeq2 and edgeR) we evaluated the differential expression between the two stages. Each method generated a p-value distribution with a large proportion of high values (and somewhat oddly shaped for DESeq2). The total number of significant genes is much higher than microarray findings, but results are similar between the two methods. With the microarray data there is a higher frequency of 'low' p-values, but the lowest value is high in comparison to p-values generated from IonTorrent. This explains the larger number of significant genes at the same FDR.       

```{r DE, fig.align='center', message=FALSE, echo=FALSE, results='asis', warning=FALSE, fig.height=9, fig.width=12}

# Set thresholds 
p.cutoff <- 0.05
fc <- 2

# DESeq2
dds <- DESeqDataSetFromMatrix(countData = seq.data, colData = seq.meta, design = ~ Donors + Stage)
dds <- DESeq(dds)
res <- results(dds) 
res$threshold <- as.logical(abs(res$log2FoldChange) > fc & res$padj < p.cutoff)

# edgeR
yv <- DGEList(counts=seq.data, group=seq.meta$Stage)
yv <- calcNormFactors(yv, method="TMM")
mod  <- model.matrix(~Stage + Donors, seq.meta)

# calculate dispersion
dge <- estimateGLMCommonDisp(yv, mod)
dge <- estimateGLMTrendedDisp(dge, mod)
dge <- estimateGLMTagwiseDisp(dge, mod)
fit <- glmFit(dge, mod)

etable <- glmLRT(fit, coef="StageP0")
res_edgeR <- topTags(etable, n=nrow(yv))$table
res_edgeR$threshold <- as.logical(abs(res_edgeR$logFC) > fc & res_edgeR$FDR < p.cutoff)

# Compare p-value distribution
par(mfrow=c(2,2))
hist(res$pvalue, col="darkgrey", border=F, xlab="P-value", main="DESeq2")
hist(res_edgeR$PValue, col="darkgrey", border=F, xlab="P-value", main="edgeR")

# Compare FC versus abundance
with(res, plot(log2(baseMean), log2FoldChange, pch=20, main="DESeq2: Fold change vs abundance"))
with(subset(res, padj< p.cutoff), points(log2(baseMean), log2FoldChange, pch=20, col="red"))
abline(h=c(-2,2), col="blue")

with(res_edgeR, plot(logCPM, logFC, pch=20, main="edgeR: Fold change vs abundance"))
with(subset(res_edgeR, FDR< p.cutoff), points(logCPM, logFC, pch=20, col="red"))
abline(h=c(-2,2), col="blue")

sigDE <- c(length(which(res$threshold)), length(which(res_edgeR$threshold)), 
           length(which(row.names(res)[which(res$threshold)] %in% row.names(res_edgeR)[which(res_edgeR$threshold)])))
methods <- c("DESeq2", "edgeR", "Overlapping")
sigDE <- data.frame(methods, "Genes"=sigDE)
colnames(sigDE)[1] <- ""
kable(sigDE, format='markdown', row.names=F)
```

### RNA-Seq versus Microarray
Fold changes are much higher for RNA-Seq data compared to microarray. A good majority of the genes identified as significant ( FDR < 0.05) using DESeq2 have fold change < 1 in the microrarray data.
```{r comparison, fig.align='center'}
# take only probes with unique ensembl identifier
mtable <- rma.data[which(fData(rma.data)$Ensembl %in% row.names(res)),]
mtable <- mtable[which(!duplicated(fData(mtable)$Ensembl)),]
res.table <- gene_list[row.names(exprs(mtable)),]

# merge array and deseq2 results
m <- match(fData(mtable)$Ensembl, row.names(res))
merged <- merge(res.table, res[m,], by.x='Ensembl', by.y='row.names')

# plot
with(merged, plot(logFC, log2FoldChange, xlab="logFC Array", ylab="logFC DESeq", pch=20, 
                  col="black", main="Array vs IonTorrent"))
with(subset(merged, padj< 0.05),  points(logFC, log2FoldChange, xlab="logFC Array", 
                                       ylab="logFC DESeq", pch=20, col="red"))
with(subset(merged, adj.P.Val < 0.1), points(logFC, log2FoldChange, xlab="logFC Array", 
                                       ylab="logFC DESeq", pch=20, col="green"))
legend("topleft", xjust=1, yjust=1, legend=c("FDR<0.05 DESeq2", "FDR<0.1 Array", "FDR>0.05"), 
       pch=20, col=c("red", "green", "black"), bty="n")
 
```


```{r compcodeR, eval=FALSE, echo=FALSE}
# compcodeR
require(compcodeR)
info.parameters <- list(dataset = "armant_data", uID = "123456")
colnames(seq.meta)[2] <- 'condition' 
cpd <- compData(count.matrix = seq.data, sample.annotations = seq.meta, info.parameters = info.parameters)
check_compData(cpd)
saveRDS(cpd, "saveddata.rds")

# run diff exp
runDiffExp(data.file = "saveddata.rds", result.extent = "voom.limma", Rmdfunction = "voom.limma.createRmd", 
           output.directory = "./compcodeR/", norm.method = "TMM")
runDiffExp(data.file = "saveddata.rds", result.extent = "edgeR.exact", Rmdfunction = "edgeR.exact.createRmd", 
           output.directory = "./compcodeR/", norm.method = "TMM", trend.method = "movingave", disp.type = "tagwise")
runDiffExp(data.file = "saveddata.rds", result.extent = "ttest", Rmdfunction = "ttest.createRmd", 
           output.directory = "./compcodeR/", norm.method = "TMM")

# comparison
file.table <- data.frame(input.files = c("./compcodeR//saveddata_edgeR.exact.rds",
                                         "./compcodeR//saveddata_ttest.rds", 
                                         "./compcodeR//saveddata_voom.limma.rds"), stringsAsFactors = FALSE)

parameters <- list(incl.nbr.samples = NULL, incl.replicates = NULL, incl.dataset = "saveddata", incl.de.methods = NULL,
                   fdr.threshold = 0.05, tpr.threshold = 0.05, typeI.threshold = 0.05, ma.threshold = 0.05, 
                   fdc.maxvar = 1500, overlap.threshold = 0.05, fracsign.threshold = 0.05, 
                   comparisons = c("auc", "fdr", "tpr", "ma", "correlation"))

runComparison(file.table = file.table, parameters = parameters, output.directory = "./compcodeR/")


```

```{r voom, echo=FALSE, eval=FALSE}
# voom limma
mod  <- model.matrix(~Stage + Donors, seq.meta)
v <- voom(dge, mod, plot=FALSE)

fit <- lmFit(v, mod)
fit <- eBayes(fit)
res_voom <- topTable(fit, coef=2, number=nrow(seq.data), sort.by="p")
res_voom$threshold <- as.logical(abs(res_voom$logFC) > fc & res_voom$adj.P.Val < p.cutoff)
```


```{r heatmap, fig.align='center', results='asis', eval=FALSE, echo=FALSE}

# Matrix of pseudocounts for downstream visualization
rld <- rlog(dds)

# Plot example heatmap 
ord.res <- res[order(res$padj),]
select <- row.names(ord.res)[1:255]
samples <- row.names(seq.meta)[order(rownames(seq.meta))]
heatmap.2(assay(rld)[select, samples],scale="row", col=heatcolors.2, labRow="", Colv=FALSE, Rowv=FALSE,
          trace="none", dendrogram="none", cexCol = 0.8)

m <- match(select, fData(rma.data)$Ensembl, nomatch=NA)
m <- m[!is.na(m)]
exprs <- exprs(rma.data)[m,]
samples <- row.names(pData(rma.data))[order(pData(rma.data)$Stage)]
select.array <- match(select, row.names(exprs), nomatch=NA)
heatmap.2(exprs[,samples],scale="row", col=heatcolors.2, labRow="", Colv=FALSE, Rowv=FALSE, 
          trace="none", dendrogram="none", cexCol = 0.8)

sigGenes <- gene_list[row.names(exprs)[1:10],]
sigGenes$P.Value <- formatC(sigGenes$P.Value, digits=3, format="e")
kable(sigGenes, format='markdown', row.names=F)
```


 
 
